pkg/ipalloc/ipalloc.go

```diff
@@ -0,0 +1,617 @@
+			size: addrDiff(left.from, right.to),
+		},
+		prev: leftOfLeft,
+		next: rightOfRight,
+	}
+
+	if leftOfLeft != nil {
+		leftOfLeft.next = &merged
+	}
+	if rightOfRight != nil {
+		rightOfRight.prev = &merged
+	}
+
+	leftIdx := slices.Index(*abl, left)
+	*abl = slices.Delete(*abl, leftIdx, leftIdx+1)
+
+	rightIdx := slices.Index(*abl, right)
+	*abl = slices.Delete(*abl, rightIdx, rightIdx+1)
+
+	*abl = append(*abl, &merged)
+
+	sort.Sort(abl)
+}
+
+// insert a new block for a given `newIP` between `left` and `right` and patch the link pointers
+func (abl *availableBlockList) insert(left, right *linkedBlock, newIP netip.Addr) {
+	block := &linkedBlock{
+		block: block{
+			from: newIP,
+			to:   newIP,
+			size: uint128{lo: 1},
+		},
+		prev: left,
+		next: right,
+	}
+	if left != nil {
+		left.next = block
+	}
+	if right != nil {
+		right.prev = block
+	}
+
+	*abl = append(*abl, block)
+}
+
+type block struct {
+	from netip.Addr
+	to   netip.Addr
+	size uint128
+}
+
+func (b *block) within(ip netip.Addr) bool {
+	return ip.Compare(b.from) >= 0 && ip.Compare(b.to) <= 0
+}
+
+func (b *block) less(other *block) bool {
+	return b.getSize().less(other.getSize())
+}
+
+func (b *block) bigIntSize() *big.Int {
+	from := big.NewInt(0)
+	from = from.SetBytes(b.from.AsSlice())
+
+	to := big.NewInt(0)
+	to = to.SetBytes(b.to.AsSlice())
+
+	return to.Sub(to, from)
+}
+
+func (b *block) calcSize() {
+	b.size = addrDiff(b.from, b.to)
+}
+
+func (b *block) getSize() uint128 {
+	// Size is stored and not calculated since its more performant during sorting where the size
+	// of the same block is frequently requested.
+	return b.size
+}
+
+type linkedBlock struct {
+	block
+
+	prev *linkedBlock
+	next *linkedBlock
+}
+
+func (lb linkedBlock) String() string {
+	var sb strings.Builder
+	if lb.prev != nil {
+		fmt.Fprintf(&sb, "(%s - %s) <- ", lb.prev.from, lb.prev.to)
+	}
+	fmt.Fprintf(&sb, "(%s - %s)[%s]", lb.from, lb.to, lb.bigIntSize().String())
+	if lb.next != nil {
+		fmt.Fprintf(&sb, " -> (%s - %s)", lb.next.from, lb.next.to)
+	}
+	return sb.String()
+}
+
+// uint128 is inspired by the unexported variant in netip. Used in our case not to represent the IPs
+// themselves but the size of blocks for sorting purposes.
+type uint128 struct {
+	hi uint64
+	lo uint64
+}
+
+func addrDiff(from, to netip.Addr) uint128 {
+	fromInt := uint128FromAddr(from)
+	toInt := uint128FromAddr(to)
+	return toInt.sub(fromInt)
+}
+
+func (u uint128) less(other uint128) bool {
+	if u.hi == other.hi {
+		return u.lo < other.lo
+	}
+
+	return u.hi < other.hi
+}
+
+func (u uint128) sub(other uint128) uint128 {
+	return uint128{
+		hi: u.hi - other.hi,
+		lo: u.lo - other.lo,
+	}
+}
+
+// uint128FromAddr extracts the uint128 value using a reflection hack.
+func uint128FromAddr(addr netip.Addr) uint128 {
+	addrStruct := reflect.ValueOf(addr).FieldByName("addr")
+	return uint128{
+		hi: addrStruct.FieldByName("hi").Uint(),
+		lo: addrStruct.FieldByName("lo").Uint(),
+	}
+}
```

Comment:
Would it be more performant and maintainable to use the `encoding/binary` pkg? For example:
```go
func uint128FromAddr(addr netip.Addr) uint128 {
	bytes := addr.As16()
	return uint128{
		hi: binary.BigEndian.Uint64(bytes[:8]),
		lo: binary.BigEndian.Uint64(bytes[8:]),
	}
}
```

Reply:
Yes! Did a quick benchmark, this goes from `167.0 ns/op` to `7.397 ns/op`, did not expect that, thanks!